virtio-queue/src/queue.rs

```diff
@@ -269,6 +269,314 @@ impl Queue {
     }
 }
 
+fn write_elem<M: GuestMemory>(
+    mem: &M,
+    addr: GuestAddress,
+    head_index: u16,
+    len: u32,
+) -> Result<(), Error> {
+    mem.write_obj(VirtqUsedElem::new(head_index.into(), len), addr)
+        .map_err(Error::GuestMemory)
+}
+
+fn store_elem<M: GuestMemory>(
+    mem: &M,
+    used_ring: GuestAddress,
+    next_used: u16,
+) -> Result<(), Error> {
+    let addr = used_ring
+        .checked_add(2)
+        .ok_or(Error::AddressOverflow)?;
+    mem.store(u16::to_le(next_used), addr, Ordering::Release)
+        .map_err(Error::GuestMemory)
+}
+
+#[cfg(kani)]
+#[allow(dead_code)]
+mod verification {
+    use std::mem::ManuallyDrop;
+    use vm_memory::MmapRegion;
+
+    use std::num::Wrapping;
+    use vm_memory::FileOffset;
+
+    use vm_memory::{GuestRegionMmap, GuestMemoryRegion, MemoryRegionAddress, AtomicAccess, GuestMemory, GuestMemoryError};
+
+    use super::*;
+
+    /// A made-for-kani version of `vm_memory::GuestMemoryMmap`. Unlike the real
+    /// `GuestMemoryMmap`, which manages a list of regions and then does a binary
+    /// search to determine which region a specific read or write request goes to,
+    /// this only uses a single region. Eliminating this binary search significantly
+    /// speeds up all queue proofs, because it eliminates the only loop contained herein,
+    /// meaning we can use `kani::unwind(0)` instead of `kani::unwind(2)`. Functionally,
+    /// it works identically to `GuestMemoryMmap` with only a single contained region.
+    pub struct ProofGuestMemory {
+        the_region: vm_memory::GuestRegionMmap,
+    }
+
+    impl GuestMemory for ProofGuestMemory {
+        type R = vm_memory::GuestRegionMmap;
```

Comment:
I replaced this with something like `type R = MyOptimizedRegion;` and then this leaded me to implement `Bytes<MemoryRegionAddress>` and `GuestMemoryRegion` for `MyOptimizedRegion`. I also struggled to implemented `new()`, which looks like:
```rust
    impl MyOptimizedRegion {
        pub fn new(start: GuestAddress, len: u64) -> Self {
            // Allocate the memory region and bitmap
            Self {
                region_start: start,
                region_len: len,
                data: vec![0; len as usize],
                region_bitmap: AtomicBitmap::new(len as usize, NonZero::new(4096).unwrap()),
            }
        }
    }
```
I am worried that this approach would add a lot of boiler plate code and thus code duplication. Any comments @roypat ? Thanks

Reply:
Can you just use `()` as the bitmap type? That'll turn all the bitmap stuff into noops, pretty sure that's what we're doing over in Firecracker. For the memory allocation, try just doing `std::alloc::alloc` (or kani::any_array() with mem::forget). No need to ever free memory, because the concept of a memory leak doesnt exist in kani